File: application-ver2/app/src/main/java/com/example/heartbeat/ECG/FragmentECG.java
package com.example.heartbeat.ECG;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.heartbeat.Command;
import com.example.heartbeat.MenuActivity;
import com.example.heartbeat.R;

public class FragmentECG extends Fragment{
    Button start;
    Button pause;
    private int count = -1;

    Command MyCmd;
    FrameLayout frameLayout;
    TextView sensorField;
    View view;
    RealTimeGraphECG myGraph;
    Thread realTimeThread;
    Activity activity;
    Boolean threadFlag;

    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragmentecg, container, false);

        start = (Button)view.findViewById(R.id.start);
        pause = (Button)view.findViewById(R.id.pause);
        sensorField = (TextView)view.findViewById(R.id.ecg_rtor_value);
        myGraph = new RealTimeGraphECG(view);

        threadFlag = false;

        start.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v){
                ((MenuActivity)getActivity()).setViewField(view, "ecg");
                ((MenuActivity)getActivity()).sendStrCmd(MyCmd.str_readecg2);
                start.setVisibility(View.GONE);
                pause.setVisibility(View.VISIBLE);
                threadFlag = true;
                realTimeStart();
            }
        });

        pause.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v){
                ((MenuActivity)getActivity()).setViewField(view, "stop");
                ((MenuActivity)getActivity()).sendStrCmd(MyCmd.str_stop);
                start.setVisibility(View.VISIBLE);
                pause.setVisibility(View.GONE);
                threadFlag = false;
                realTimeThread.interrupt();
            }
        });

        LayoutInflater layoutInflater = (LayoutInflater)getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        frameLayout = ((MenuActivity)getActivity()).getFrameLayout();

        MyCmd = new Command();

        return view;
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (context instanceof Activity)
            activity = (Activity) context;
    }

    @Override
    public void onDetach() {
        ((MenuActivity)getActivity()).sendStrCmd(MyCmd.str_stop);
        ((MenuActivity)getActivity()).mode="stop";
        frameLayout.setVisibility(View.GONE);
        if(realTimeThread != null)realTimeThread.interrupt();
        super.onDetach();

    }



    public void realTimeStart(){
        realTimeThread = new Thread(new Runnable() {
            @Override
            public void run() {
                while(threadFlag){
                    try {
                        Thread.sleep(100);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            int tmp_count;
                            tmp_count = ((MenuActivity)activity).ecgData.getCount();
                            //Log.i("count : ", Integer.toString(tmp));

                            //if((tmp%35) == 0)myGraph.resetMaxandMin();
                            if(count != tmp_count) {
                                Log.i("그래프 그리기 메서드" , "호출!");
                                count = tmp_count;
                                myGraph.addEntry(((MenuActivity) activity).valueForGraph);
                            }
                        }
                    });
                }
            }
        });

        realTimeThread.start();
    }


}

File: Application_part/project_kotlin_version/Kotlin_ver/app/src/main/java/com/example/kotlin_ver/ECG/FragmentECG.kt
package com.example.kotlin_ver.ECG

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.example.kotlin_ver.Command
import com.example.kotlin_ver.MenuActivity
import com.example.kotlin_ver.R
import com.google.android.material.internal.ContextUtils.getActivity
import kotlin.collections.ArrayList

class FragmentECG : Fragment() {
    lateinit var start: Button
    lateinit var pause: Button
    lateinit var sensorField: TextView
    lateinit var rtorField: TextView
    lateinit var bpmField: TextView
    lateinit var myGraph: RealTimeGraphECG

    var activity: Activity? = null

    val ecgDataList: ArrayList<ECGData> = ArrayList()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view : View = inflater.inflate(R.layout.fragmentecg, container, false)
        start = view.findViewById(R.id.start)
        pause = view.findViewById(R.id.pause)
        sensorField = view.findViewById(R.id.ecg_rtor_value) as TextView
        rtorField = view.findViewById(R.id.ecg_rtor_value)
        bpmField = view.findViewById(R.id.ecg_rtorbpm_value)

        myGraph = RealTimeGraphECG(view)

        start.setOnClickListener {
            (activity as MenuActivity).sendStrCmd(Command.str_readecg2)
            start.visibility = View.GONE
            pause.visibility = View.VISIBLE
        }

        pause.setOnClickListener {
            (activity as MenuActivity).sendStrCmd(Command.str_stop)
            start.visibility = View.VISIBLE
            pause.visibility = View.GONE
        }

        return view
    }


    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is Activity) activity = context
    }

    override fun onDetach() {
        (getActivity() as MenuActivity).sendStrCmd(Command.str_stop)
        (getActivity() as MenuActivity).mode = "stop"
        (getActivity() as MenuActivity).fragmentContainerLayout.visibility = View.GONE
        super.onDetach()
    }

    fun addData(hexaData: ByteArray){
        val ecg = ECGData(hexaData)
        ecgDataList.add(ecg)

        rtorField.text = "${ecg.RtoR}"
        bpmField.text = "${ecg.RtoRBpm}"

        activity!!.runOnUiThread {
            myGraph.addEntry(ecg)
        }

    }

}